torch/csrc/utils/tensor_conversion_dispatch.cpp

```diff
@@ -14,18 +14,16 @@ namespace torch { namespace utils {
 at::Tensor dispatch_type_conversion(
     const at::Tensor& self,
     const at::Type& type,
-    c10::optional<int32_t> device_index,
+    c10::optional<c10::Device> device_opt,
```

Comment:
I'm not convinced this is a good thing to change without ripping out "type" here -- this API is now overspecified -- who wins if type.device_type() doesn't match device_opt?  new_with_tensor_copy has the same issue but there are asserts there and it's not a general API.